### PR TITLE
add the following member methods in Some / None classes :

### PR DESCRIPTION
**Pull request by alexandervanhecke**
- contains
- isEmpty
- isDefined
- nonEmpty
- forEach
- exists
- orEmpty

see also issue https://github.com/gcanti/fp-ts/issues/183

### CODE REVIEW

**Comment by alexandervanhecke** [comment]
@SimonMeskens no other reason than I got my inspiration from the scala Option type which defines these methods and aliases.  But it's definitely debatable.

wkr,

**Comment by SimonMeskens** [comment]
I figured that was the case, just wanted to check. Probably better if you didn't waste a full closure on it though, maybe just:

```TypeScript
nonEmpty: () => boolean = this.isDefined
```

It's not a big deal, since this is not performance critical code and the latest V8 optimizes specially for this case, but it seems a bit wasteful to create a full closure just for an alias. Added benefit is that without the closure, nonEmpty is strictly equal to isDefined.

**Comment by alexandervanhecke** [comment]
@SimonMeskens thanks for your feedback, I wasn't even aware this could have a performance impact.

File: src/Option.ts [hunk 1]
```diff
@@ -204,6 +225,27 @@ export class Some<A>
   toString() {
     return `some(${toString(this.value)})`
   }
+  contains(setoid: Setoid<A>, a: A): boolean {
```
Comment (SimonMeskens) [inline src/Option.ts]: Would it make more sense to make `setoid` optional and fall back to `this.value.equals(a)`?

**Comment by SimonMeskens** [comment]
> thanks for your feedback, I wasn't even aware this could have a performance impact.

If you wrap a function call inside another function, you always create a new function with a new callstack. Arrow functions don't have a full context and are a bit lighter, but just using equality like I suggested doesn't have any overhead.

As I said, it doesn't matter much for performance, but I think it's just good habit not to create closures when you don't need to. In older versions of V8, creating a new closure is about 2x slower, in newer versions, it's almost as fast in this specific case, because V8 will inline the call for you. Unless you'll be calling nonEmpty a couple thousand times per second, you won't even notice a linear increase like that.

File: src/Option.ts (hunk 1 above)
Comment (alexandervanhecke) [inline src/Option.ts]: Good point, I've tried to mirror the signature of the `equals` method, but an optional `setoid` would work as well of course.  Let me know what you prefer, and I'll update if needed.

Comment (SimonMeskens) [inline src/Option.ts]: Ah yes, I see how it mirrors `equals`. I don't like how that breaks compatibility with most other libraries. Is there a reason for that @gcanti?

**Comment by gcanti** [comment]
@alexandervanhecke `.isEmpty`, `.isDefined` and `.nonEmpty` do the same thing, can we keep just one? Or maybe two of them, with the following renaming (for consistency with the top-level functions)

`isEmpty` -> `isNone`
`isDefined` -> `isSome`

> Is there a reason for that?

@SimonMeskens yes, you have to provide an explicit `Setoid` instance, corresponding to the type constraint `Eq a =>` in the following signature

```purescript
contains :: Eq a => Maybe a -> a -> Boolean
//           ^       ^
//        setoid    this
```

This encoding, i.e. translating a type constraint to an additional required dictionary, usually in first position, is consistent across the library.

> `exists(p: (a: A) => Boolean): boolean`

I guess should be **b**oolean

> `.forEach<B>(f: (a: A) => B): void`

As is this signature means that some side effect will be executed, should be removed.

`.orEmpty` seems too specialised and a bit weird, it requires a `Monoid` instance, but if you have such an instance in scope can't you just call `.getOrElse(monoid.empty())`?

**Comment by SimonMeskens** [comment]
Yes, I do understand the consistency, I just don't like the fact that fp-ts is incompatible with fantasy-land in so many places, and thus incompatible with libraries that consume fantasy types. Maybe we should write an overload for every compatibility break?

**Comment by gcanti** [comment]
I see, we could discuss this topic in a new issue, though being compatible with fantasy-land and static-land is not a goal and AFAIK there are already many unavoidable incompatibilities

**Comment by alexandervanhecke** [comment]
@gcanti thanks for your feedback, I've removed these functions and did the renames as you suggested.  

as for the `isSome: () => boolean = () => true` syntax, I ended up not using it because it seemed to trigger errors in some of the other mocha tests

let me know if there are other things you want me to change

**Comment by gcanti** [comment]
Thanks @alexandervanhecke, let me merge some other PRs and then I'll publish a new release with this

**Pull request merged**